Unix usocket settimeout fix

eladnam opened the pull request:
fix socket_settimeout function in unix port. 
old impl fails on s.settimeout(None)

eladnam committed:
unix/modusocket: fix socket_settimeout implementaion.

The original code called setsockopt(SO_RCVTIMEO/SO_SNDTIMEO) with
null timeout structure argument, which is an illegal usage of that function.
The old code also didn't validate the return value of setsockopt,
missing the whole bug completely.

dpgeorge commented:
Thanks for the patch.  But can you please explain exactly how the old implementation was failing, what went wrong and why this patch fixes it?

pfalcon commented:
It probably makes sense to squash all the patches into one.

Then, the original code thought that the way to cancel POSIX timeouts is to call setsockopt(SO_RCVTIMEO/SO_SNDTIMEO) with NULL timeout structure argument. Which is quite logical, that's apparently why the original author of https://github.com/micropython/micropython/pull/2381 did it like that, and why the maintainer (me) didn't look at it critically (despite spotting other issues with that PR, which needed rework). But both man "man 7 socket" on Linux, and POSIX itself (http://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html) are "kind explicit" that zero value of timeout means no timeout, and silent about passing NULL as a timeout, so the only conclusion that indeed cancelling a timeout requires setting zero timeout.

All this should be described in the commit message (with references, like above). And given that it's still confusing, there should be code comment too (e.g. before `if (timeout_in != mp_const_none)`, something like "Timeout of None means no (== infinite) timeout, which in POSIX is signified with 0 timeout, and that's how 'tv' is initialized above").

eladnam commented:
> Thanks for the patch. But can you please explain exactly how the old implementation was failing, what went wrong and why this patch fixes it?

as @pfalcon explained, the old code pass NULL as optval, which is illegal usage of setsockopt.
the old code also didn't validate the return value of the function, missing the failure.
i've added explanation to the commit message.



> It probably makes sense to squash all the patches into one.
> 
> Then, the original code thought that the way to cancel POSIX timeouts is to call setsockopt(SO_RCVTIMEO/SO_SNDTIMEO) with NULL timeout structure argument. Which is quite logical, that's apparently why the original author of #2381 did it like that, and why the maintainer (me) didn't look at it critically (despite spotting other issues with that PR, which needed rework). But both man "man 7 socket" on Linux, and POSIX itself (http://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html) are "kind explicit" that zero value of timeout means no timeout, and silent about passing NULL as a timeout, so the only conclusion that indeed cancelling a timeout requires setting zero timeout.
> 
> All this should be described in the commit message (with references, like above). And given that it's still confusing, there should be code comment too (e.g. before `if (timeout_in != mp_const_none)`, something like "Timeout of None means no (== infinite) timeout, which in POSIX is signified with 0 timeout, and that's how 'tv' is initialized above").

ok, i've squashed the commits and added a comment to the code.

dpgeorge commented:
Ok, thanks for the explanation, sounds good to me.

Merged in 3f54462adddb3210b46f3dba3f1879cd01e2f16e (with minor formatting change to put call to `setsockopt` on one line).

pfalcon commented:
The explanation however wasn't included in the commit message, as was suggested...
